src/pygeoprocessing/geoprocessing.py

```diff
@@ -1831,14 +1836,28 @@ def reclassify_raster(
     """
     if len(value_map) == 0:
         raise ValueError("value_map must contain at least one value")
+    if target_nodata is None:
+        raise ValueError("target_nodata must be a valid number, received None")
```

Comment:
I _think_ there could be an argument for having `target_nodata=None` (though I don't know if we use this in InVEST).  It seems to me that the main case we're trying to prevent is a broken reclassification rule ... if `base_nodata` is defined but `target_nodata` is not, then how do we translate that?  In that case, I absolutely think we should raise an exception.

Reply:
Yea, I think you're right in that for this case the function does not have enough information on what to do, so should raise an exception. I was coming from the mindset that `target_nodata` should not be allowed to be None at all, which would avoid this situation. I'd love to hear the argument for `target_nodata=None` as being something a user would explicitly want. I guess I'd prefer to avoid over engineering if not allowing `target_nodata` to be None doesn't restrict the use of the function in anyway...

Reply:
Looking through the InVEST source code, the only concrete example of this that I found is that GLOBIO allows a user-defined GLOBIO LULC, and we use the nodata value, whatever it is (even if it's `None`).

I also looked through the first 10 pages of a github search for `reclassify_raster` and the GLOBIO case is the only one I can find where the nodata value _could_ be set.  So I'll concede that this is probably an edge case.

But what makes me wary of requiring a nodata value is that it's [an explicitly optional part of the GDAL raster data model](https://gdal.org/user/raster_data_model.html#raster-band).  That's not to say that we have to support everything that GDAL supports ... GDAL rasters support mask bands, for instance, which we've never used to my knowledge.

I still think that the best example is not so much a matter of explicitly wanting a nodata value as much as it is that a nodata value might not be needed.  If a user doesn't need a nodata value, why should one be required?

Reply:
I believe I addressed this here: https://github.com/natcap/pygeoprocessing/issues/220

Allowing target nodata values of None as long as they don't break the case you mention above (target_nodata = None and base raster nodata not present in value_map)